### PR TITLE
parallel layer

### PR DESCRIPTION
**Pull request by adam-r-kowalski**
This is aimed to simply the case when you wish to call multiple layers in parallel rather than in sequence. For example in the resnet style approach you apply a layer and the identity function then sum them. This can be written like so

```
model = Chain(Parallel(Conv((3, 3), 3=>3, pad=1),
                       x -> x),
              sum)

inputs = randn(28, 28, 3, 1)
model(inputs)
```

If you wish to have a set of convolutional layers that can extract different sized features and then concat them together of the feature dimension.

```
model = Chain(Parallel(Conv((3, 3), 3=>3),
                       Conv((5, 5), 3=>3, pad=1),
                       Conv((7, 7), 3=>3, pad=2)),
              x -> cat(x..., dims=3))


inputs = randn(28, 28, 3, 1)
model(inputs)
```

### CODE REVIEW

**Comment by jpsamaroo** [comment]
I'm guessing this layer is similar to the one implemented in https://github.com/FluxML/Flux.jl/pull/662? If so, maybe @swiesend can provide a review.

**Comment by adam-r-kowalski** [comment]
I think this is more similar to what Jax does. [here is there resnet implementation](https://github.com/google/jax/blob/master/examples/resnet50.py#L38).

Currently my layer doesn't actually do anything in parallel, it's more about the semantics of saying I would like to apply all these layers to my input and then do something with the result such as `sum` `reduce` `cat` etc rather then doing them in a sequential fashion.

**Comment by jpsamaroo** [comment]
> Currently my layer doesn't actually do anything in parallel

AFAICT neither does #662, which I definitely think is a good thing, since adding multiple threads/processes/etc. into the mix can make things a lot more complicated. If we want actual parallelism, I think that should be done as a separate PR so as to not block this very useful functionality being added to Flux.

**Comment by jpsamaroo** [comment]
Also, I think we'll need a few tests for this layer. Maybe you can draw inspiration from #662, if things are similar enough.

**Comment by swiesend** [comment]
I like the `Tuple` approach from this #773.

The thing with #662 is, that it should maybe be better called `MapReduce` instead of `Parallel`, as both phases are fused into one layer. It is possible to do nothing in the reduce phase, but only by applying the `identity` function, which is quite silly.

**Comment by swiesend** [comment]
The only concern I have is, that a thing like this should also be able to work smoothly for sequential models, to overcome the border between the convolutional and sequential models... or is there none?

**Comment by adam-r-kowalski** [comment]
@jpsamaroo I have added some tests based on tests for the `Chain` layer. It is honestly almost identical to that layer. The only difference is that instead of calling in sequence, I do a list comprehension and call each layer on the input.

**Comment by HenriDeh** [comment]
This does not seem to be compatible with `Flux.gpu`:

```
julia> m = Chain(Dense(10,5), Parallel(Dense(5,5), Dense(5,1))) |> gpu
julia> m(cu(rand(10)))
ERROR: ArgumentError: cannot take the CPU address of a CuArray{Float32,1,Nothing}
Stacktrace:
 [1] unsafe_convert(::Type{Ptr{Float32}}, ::CuArray{Float32,1,Nothing}) at /home/dehaybe/.julia/packages/CuArrays/ZYCpV/src/array.jl:212
 [2] gemv!(::Char, ::Bool, ::Array{Float32,2}, ::CuArray{Float32,1,Nothing}, ::Bool, ::CuArray{Float32,1,Nothing}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/blas.jl:587
 [3] gemv!(::CuArray{Float32,1,Nothing}, ::Char, ::Array{Float32,2}, ::CuArray{Float32,1,Nothing}, ::Bool, ::Bool) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/matmul.jl:463
 [4] * at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/matmul.jl:66 [inlined]
 [5] (::Dense{typeof(σ),Array{Float32,2},Array{Float32,1}})(::CuArray{Float32,1,Nothing}) at /home/dehaybe/.julia/packages/Flux/oX9Pi/src/layers/basic.jl:102
 [6] Dense at /home/dehaybe/.julia/packages/Flux/oX9Pi/src/layers/basic.jl:113 [inlined]
 [7] #50 at ./none:0 [inlined]
 [8] iterate at ./generator.jl:47 [inlined]
 [9] collect(::Base.Generator{Tuple{Dense{typeof(σ),Array{Float32,2},Array{Float32,1}},Dense{typeof(elu),Array{Float32,2},Array{Float32,1}}},var"#50#51"{CuArray{Float32,1,Nothing}}}) at ./array.jl:622
 [10] (::Chain{Tuple{Dense{typeof(elu),CuArray{Float32,2,Nothing},CuArray{Float32,1,Nothing}},Parallel{Tuple{Dense{typeof(σ),Array{Float32,2},Array{Float32,1}},Dense{typeof(elu),Array{Float32,2},Array{Float32,1}}}}}})(::CuArray{Float32,1,Nothing}) at /home/dehaybe/Documents/Doctorat/Code/Julia/Stochastic Inventory Problem/src/custom_layers.jl:27
 [11] top-level scope at none:0
```

The faulty line appears to be `(p::Parallel)(x) = [layer(x) for layer in p.layers]` which creates an array on the cpu.

File: src/layers/basic.jl [hunk 1]
```diff
@@ -59,6 +59,46 @@ function activations(c::Chain, input)
...
+struct Parallel{T<:Tuple}
+  layers::T
+  Parallel(xs...) = new{typeof(xs)}(xs)
+end
+
+@forward Parallel.layers Base.getindex, Base.length, Base.first, Base.last,
+  Base.iterate, Base.lastindex
+
+children(p::Parallel) = p.layers
+mapchildren(f, p::Parallel) = Parallel(f.(c.layers)...)
```
Comment (HenriDeh) [inline src/layers/basic.jl]: c.layers should be p.layers

**Comment by ToucheSir** [comment]
We now have https://github.com/FluxML/Flux.jl/pull/1462, so this can be closed.

**Pull request closed without merging**